# 01. Importing libraries

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

# 02. Importing data

In [2]:
#Define path
path = r'C:\Users\Josh Abrams\Desktop\09-2022 Instacart Basket Analysis'

In [3]:
df_ords_prods_merge= pd.read_pickle(os.path.join(path, 'Data', 'Prepared Data', 'orders_products_merged_update.pkl'))


In [4]:
df_ords_prods_merge= pd.read_pickle(r'C:\Users\Josh Abrams\Desktop\09-2022 Instacart Basket Analysis\Data\Prepared Data\orders_products_merged_update.pkl')

In [5]:
df = df_ords_prods_merge

In [6]:
df.head()

Unnamed: 0_x  order_id  customer_id eval_set  order_number  \
0             0   2539329            1    prior             1   
1             1   2398795            1    prior             2   
2             2    473747            1    prior             3   
3             3   2254736            1    prior             4   
4             4    431534            1    prior             5   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  product_id  \
0                   2                  8                     NaN         196   
1                   3                  7                    15.0         196   
2                   3                 12                    21.0         196   
3                   4                  7                    29.0         196   
4                   4                 15                    28.0         196   

   add_to_cart_order  ...  Unnamed: 0_y  product_name aisle_id  department_id  \
0                  1  ...           195          Soda       77              7   
1                  1  ...           195          Soda       77              7   
2                  1  ...           195          Soda       77              7   
3                  1  ...           195          Soda       77              7   
4                  1  ...           195          Soda       77              7   

   prices  _merge    price_range_loc     busiest_day        busiest_days  \
0     9.0    both  Mid-range product  Regularly busy  Regulary busy days   
1     9.0    both  Mid-range product  Regularly busy        Slowest days   
2     9.0    both  Mid-range product  Regularly busy        Slowest days   
3     9.0    both  Mid-range product      Least busy        Slowest days   
4     9.0    both  Mid-range product      Least busy        Slowest days   

  busiest_period_of_day  
0           Most orders  
1        Average orders  
2           Most orders  
3        Average orders  
4           Most orders  

[5 rows x 21 columns]

In [7]:
df.shape

(32404859, 21)

In [8]:
# Find the mean number of orders by department id
df_ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## We now have the mean number of orders for all department ids.  Not just ones that appear in the subset of the first 1,000,000 rows.  There is now a greater range between the department with the lowest average number of orders (dept id 5 at 15.46) and the department with the highest (dept id 21 at 22.90)

## Question 4&5

In [9]:
# Creating a loyalty flag for existing customers using the transform() and loc() functions.
# Step one: Create new column with max order
df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['customer_id'])['order_number'].transform(np.max)

In [10]:
df_ords_prods_merge.head(15)

Unnamed: 0_x  order_id  customer_id eval_set  order_number  \
0              0   2539329            1    prior             1   
1              1   2398795            1    prior             2   
2              2    473747            1    prior             3   
3              3   2254736            1    prior             4   
4              4    431534            1    prior             5   
5              5   3367565            1    prior             6   
6              6    550135            1    prior             7   
7              7   3108588            1    prior             8   
8              8   2295261            1    prior             9   
9              9   2550362            1    prior            10   
10           144   2968173           15    prior            15   
11           146   1870022           15    prior            17   
12           147   1911383           15    prior            18   
13           150   2715276           15    prior            21   
14           151    487368           15    prior            22   

    orders_day_of_week  order_hour_of_day  days_since_prior_order  product_id  \
0                    2                  8                     NaN         196   
1                    3                  7                    15.0         196   
2                    3                 12                    21.0         196   
3                    4                  7                    29.0         196   
4                    4                 15                    28.0         196   
5                    2                  7                    19.0         196   
6                    1                  9                    20.0         196   
7                    1                 14                    14.0         196   
8                    1                 16                     0.0         196   
9                    4                  8                    30.0         196   
10                   1                  9                     7.0         196   
11                   2                 16                     8.0         196   
12                   2                 11                     7.0         196   
13                   1                  9                     7.0         196   
14                   1                 10                    14.0         196   

    add_to_cart_order  ...  product_name  aisle_id department_id  prices  \
0                   1  ...          Soda        77             7     9.0   
1                   1  ...          Soda        77             7     9.0   
2                   1  ...          Soda        77             7     9.0   
3                   1  ...          Soda        77             7     9.0   
4                   1  ...          Soda        77             7     9.0   
5                   1  ...          Soda        77             7     9.0   
6                   1  ...          Soda        77             7     9.0   
7                   2  ...          Soda        77             7     9.0   
8                   4  ...          Soda        77             7     9.0   
9                   1  ...          Soda        77             7     9.0   
10                  2  ...          Soda        77             7     9.0   
11                  6  ...          Soda        77             7     9.0   
12                  1  ...          Soda        77             7     9.0   
13                  1  ...          Soda        77             7     9.0   
14                  1  ...          Soda        77             7     9.0   

    _merge    price_range_loc     busiest_day        busiest_days  \
0     both  Mid-range product  Regularly busy  Regulary busy days   
1     both  Mid-range product  Regularly busy        Slowest days   
2     both  Mid-range product  Regularly busy        Slowest days   
3     both  Mid-range product      Least busy        Slowest days   
4     both  Mid-range product      Least busy        Slowest days   
5     both  Mid-ran

In [11]:
# Step two: Create loyalty flag column based on max order criteria
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
df_ords_prods_merge.loc[(df_ords_prods_merge['max_order'] <= 40) & (df_ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [12]:
df_ords_prods_merge.shape

(32404859, 23)

In [13]:
df_ords_prods_merge.head(15)

Unnamed: 0_x  order_id  customer_id eval_set  order_number  \
0              0   2539329            1    prior             1   
1              1   2398795            1    prior             2   
2              2    473747            1    prior             3   
3              3   2254736            1    prior             4   
4              4    431534            1    prior             5   
5              5   3367565            1    prior             6   
6              6    550135            1    prior             7   
7              7   3108588            1    prior             8   
8              8   2295261            1    prior             9   
9              9   2550362            1    prior            10   
10           144   2968173           15    prior            15   
11           146   1870022           15    prior            17   
12           147   1911383           15    prior            18   
13           150   2715276           15    prior            21   
14           151    487368           15    prior            22   

    orders_day_of_week  order_hour_of_day  days_since_prior_order  product_id  \
0                    2                  8                     NaN         196   
1                    3                  7                    15.0         196   
2                    3                 12                    21.0         196   
3                    4                  7                    29.0         196   
4                    4                 15                    28.0         196   
5                    2                  7                    19.0         196   
6                    1                  9                    20.0         196   
7                    1                 14                    14.0         196   
8                    1                 16                     0.0         196   
9                    4                  8                    30.0         196   
10                   1                  9                     7.0         196   
11                   2                 16                     8.0         196   
12                   2                 11                     7.0         196   
13                   1                  9                     7.0         196   
14                   1                 10                    14.0         196   

    add_to_cart_order  ...  aisle_id  department_id prices  _merge  \
0                   1  ...        77              7    9.0    both   
1                   1  ...        77              7    9.0    both   
2                   1  ...        77              7    9.0    both   
3                   1  ...        77              7    9.0    both   
4                   1  ...        77              7    9.0    both   
5                   1  ...        77              7    9.0    both   
6                   1  ...        77              7    9.0    both   
7                   2  ...        77              7    9.0    both   
8                   4  ...        77              7    9.0    both   
9                   1  ...        77              7    9.0    both   
10                  2  ...        77              7    9.0    both   
11                  6  ...        77              7    9.0    both   
12                  1  ...        77              7    9.0    both   
13                  1  ...        77              7    9.0    both   
14                  1  ...        77              7    9.0    both   

      price_range_loc     busiest_day        busiest_days  \
0   Mid-range product  Regularly busy  Regulary busy days   
1   Mid-range product  Regularly busy        Slowest days   
2   Mid-range product  Regularly busy        Slowest days   
3   Mid-range product      Least busy        Slowest days   
4   Mid-range product      Least busy        Slowest days   
5   Mid-range product  Regularly busy  Regulary busy days   
6   Mid-range product  Regularly busy        Busiest days   
7   Mid-range product  Regularly busy    

In [14]:
# Count number of customers by loyalty flag
df_ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [15]:
# Check basis statistics of product prices for each loyalty category
df_ords_prods_merge.groupby('loyalty_flag').agg({'prices':['mean','min','max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

## Question 6

In [16]:
# Target customers based on spending
# Step One: Create column of mean price by user
df_ords_prods_merge['mean_order'] = df_ords_prods_merge.groupby(['customer_id'])['prices'].transform(np.mean)

In [17]:
df_ords_prods_merge.shape

(32404859, 24)

In [18]:
df_ords_prods_merge.head(20)

Unnamed: 0_x  order_id  customer_id eval_set  order_number  \
0              0   2539329            1    prior             1   
1              1   2398795            1    prior             2   
2              2    473747            1    prior             3   
3              3   2254736            1    prior             4   
4              4    431534            1    prior             5   
5              5   3367565            1    prior             6   
6              6    550135            1    prior             7   
7              7   3108588            1    prior             8   
8              8   2295261            1    prior             9   
9              9   2550362            1    prior            10   
10           144   2968173           15    prior            15   
11           146   1870022           15    prior            17   
12           147   1911383           15    prior            18   
13           150   2715276           15    prior            21   
14           151    487368           15    prior            22   
15           209   2293453           19    prior             2   
16           212   1973799           19    prior             5   
17           214    532817           19    prior             7   
18           232   1573906           21    prior            10   
19           458   1593000           31    prior            10   

    orders_day_of_week  order_hour_of_day  days_since_prior_order  product_id  \
0                    2                  8                     NaN         196   
1                    3                  7                    15.0         196   
2                    3                 12                    21.0         196   
3                    4                  7                    29.0         196   
4                    4                 15                    28.0         196   
5                    2                  7                    19.0         196   
6                    1                  9                    20.0         196   
7                    1                 14                    14.0         196   
8                    1                 16                     0.0         196   
9                    4                  8                    30.0         196   
10                   1                  9                     7.0         196   
11                   2                 16                     8.0         196   
12                   2                 11                     7.0         196   
13                   1                  9                     7.0         196   
14                   1                 10                    14.0         196   
15                   5                 14                     6.0         196   
16                   6                 12                     8.0         196   
17                   4                 17                     6.0         196   
18                   3                 10                     6.0         196   
19                   3                  8                     7.0         196   

    add_to_cart_order  ...  department_id  prices _merge    price_range_loc  \
0                   1  ...              7     9.0   both  Mid-range product   
1                   1  ...              7     9.0   both  Mid-range product   
2                   1  ...              7     9.0   both  Mid-range product   
3                   1  ...              7     9.0   both  Mid-range product   
4                   1  ...              7     9.0   both  Mid-range product   
5                   1  ...              7     9.0   both  Mid-range product   
6                   1  ...              7     9.0   both  Mid-range product   
7                   2  ...              7     9.0   both  Mid-range product   
8                   4  ...              7     9.0   both  Mid-range product   
9                   1  ...              7     9.0   both  Mid-range product   
10                  2  ...              7     

In [19]:
# Step two: Assign spending flag based on average price

df_ords_prods_merge.loc[df_ords_prods_merge['mean_order']<10, 'spending_flag'] = 'Low spender'
df_ords_prods_merge.loc[df_ords_prods_merge['mean_order']>=10, 'spending_flag'] = 'High spender'

In [20]:
df_ords_prods_merge['spending_flag'].value_counts(dropna=False)

Low spender     31770614
High spender      634245
Name: spending_flag, dtype: int64

In [21]:
df_ords_prods_merge[['customer_id', 'mean_order', 'spending_flag']].head(100)

customer_id  mean_order spending_flag
0             1    6.367797   Low spender
1             1    6.367797   Low spender
2             1    6.367797   Low spender
3             1    6.367797   Low spender
4             1    6.367797   Low spender
..          ...         ...           ...
95          360   10.006250  High spender
96          377    8.496552   Low spender
97          387    7.396610   Low spender
98          420    7.387805   Low spender
99          420    7.387805   Low spender

[100 rows x 3 columns]

## Question 7

In [22]:
# Create order frequency flag based on median days since previous order.
# Step one: Create a new column showing median days since previous order.
df_ords_prods_merge['median_days_since_previous_order'] = df_ords_prods_merge.groupby(['customer_id'])['days_since_prior_order'].transform(np.median)

In [23]:
df_ords_prods_merge.shape

(32404859, 26)

In [24]:
df_ords_prods_merge.head()

Unnamed: 0_x  order_id  customer_id eval_set  order_number  \
0             0   2539329            1    prior             1   
1             1   2398795            1    prior             2   
2             2    473747            1    prior             3   
3             3   2254736            1    prior             4   
4             4    431534            1    prior             5   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  product_id  \
0                   2                  8                     NaN         196   
1                   3                  7                    15.0         196   
2                   3                 12                    21.0         196   
3                   4                  7                    29.0         196   
4                   4                 15                    28.0         196   

   add_to_cart_order  ...  _merge    price_range_loc     busiest_day  \
0                  1  ...    both  Mid-range product  Regularly busy   
1                  1  ...    both  Mid-range product  Regularly busy   
2                  1  ...    both  Mid-range product  Regularly busy   
3                  1  ...    both  Mid-range product      Least busy   
4                  1  ...    both  Mid-range product      Least busy   

         busiest_days  busiest_period_of_day  max_order  loyalty_flag  \
0  Regulary busy days            Most orders         10  New customer   
1        Slowest days         Average orders         10  New customer   
2        Slowest days            Most orders         10  New customer   
3        Slowest days         Average orders         10  New customer   
4        Slowest days            Most orders         10  New customer   

  mean_order spending_flag median_days_since_previous_order  
0   6.367797   Low spender                             20.5  
1   6.367797   Low spender                             20.5  
2   6.367797   Low spender                             20.5  
3   6.367797   Low spender                             20.5  
4   6.367797   Low spender                             20.5  

[5 rows x 26 columns]

In [25]:
# Step two: Assign frequency flag based on median days since previous order
df_ords_prods_merge.loc[df_ords_prods_merge['median_days_since_previous_order']>20, 'Frequency_flag'] = 'Non-frequent customer'
df_ords_prods_merge.loc[(df_ords_prods_merge['median_days_since_previous_order']>10) & (df_ords_prods_merge['median_days_since_previous_order']<=20), 'Frequency_flag'] = 'Regular customer'
df_ords_prods_merge.loc[df_ords_prods_merge['median_days_since_previous_order']<=10, 'Frequency_flag'] = 'Frequent customer'

In [26]:
df_ords_prods_merge.shape

(32404859, 27)

In [27]:
df_ords_prods_merge['Frequency_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: Frequency_flag, dtype: int64

In [28]:
df_ords_prods_merge[['customer_id', 'median_days_since_previous_order', 'Frequency_flag']].head(100)

customer_id  median_days_since_previous_order         Frequency_flag
0             1                              20.5  Non-frequent customer
1             1                              20.5  Non-frequent customer
2             1                              20.5  Non-frequent customer
3             1                              20.5  Non-frequent customer
4             1                              20.5  Non-frequent customer
..          ...                               ...                    ...
95          360                               4.0      Frequent customer
96          377                              16.5       Regular customer
97          387                               8.0      Frequent customer
98          420                               7.0      Frequent customer
99          420                               7.0      Frequent customer

[100 rows x 3 columns]

In [29]:
# Renaming mean_order column to mean_price to make it easier to understand
df_ords_prods_merge.rename(columns = {'mean_order':'mean_price'}, inplace = True)

In [30]:
df_ords_prods_merge.head()

Unnamed: 0_x  order_id  customer_id eval_set  order_number  \
0             0   2539329            1    prior             1   
1             1   2398795            1    prior             2   
2             2    473747            1    prior             3   
3             3   2254736            1    prior             4   
4             4    431534            1    prior             5   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  product_id  \
0                   2                  8                     NaN         196   
1                   3                  7                    15.0         196   
2                   3                 12                    21.0         196   
3                   4                  7                    29.0         196   
4                   4                 15                    28.0         196   

   add_to_cart_order  ...    price_range_loc     busiest_day  \
0                  1  ...  Mid-range product  Regularly busy   
1                  1  ...  Mid-range product  Regularly busy   
2                  1  ...  Mid-range product  Regularly busy   
3                  1  ...  Mid-range product      Least busy   
4                  1  ...  Mid-range product      Least busy   

         busiest_days  busiest_period_of_day  max_order  loyalty_flag  \
0  Regulary busy days            Most orders         10  New customer   
1        Slowest days         Average orders         10  New customer   
2        Slowest days            Most orders         10  New customer   
3        Slowest days         Average orders         10  New customer   
4        Slowest days            Most orders         10  New customer   

  mean_price spending_flag median_days_since_previous_order  \
0   6.367797   Low spender                             20.5   
1   6.367797   Low spender                             20.5   
2   6.367797   Low spender                             20.5   
3   6.367797   Low spender                             20.5   
4   6.367797   Low spender                             20.5   

          Frequency_flag  
0  Non-frequent customer  
1  Non-frequent customer  
2  Non-frequent customer  
3  Non-frequent customer  
4  Non-frequent customer  

[5 rows x 27 columns]

# CLEANING UP THE DATA FROM 4-9


In [31]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 100]

Unnamed: 0_x  order_id  customer_id eval_set  order_number  \
10030345           171    912404           17    prior            12   
10030346           181    603376           17    prior            22   
10030347          1944   3264360          135    prior             2   
10030348          1945    892534          135    prior             3   
10030349          5708    229704          342    prior             8   
...                ...       ...          ...      ...           ...   
29166209       3386542   2249946       204099    prior            29   
29166210       3386544   2363282       204099    prior            31   
29166211       3391313   3181945       204395    prior            13   
29166212       3405096   2486215       205227    prior             7   
29166213       3405099   2232716       205227    prior            10   

          orders_day_of_week  order_hour_of_day  days_since_prior_order  \
10030345                   2                 14                     5.0   
10030346                   6                 16                     4.0   
10030347                   2                 21                    13.0   
10030348                   0                  8                    12.0   
10030349                   1                 19                    30.0   
...                      ...                ...                     ...   
29166209                   0                  8                     4.0   
29166210                   0                  9                     2.0   
29166211                   3                 15                     8.0   
29166212                   3                 20                     4.0   
29166213                   0                 11                     4.0   

          product_id  add_to_cart_order  ...     price_range_loc  \
10030345       21553                  5  ...  High-range product   
10030346       21553                  3  ...  High-range product   
10030347       21553                  6  ...  High-range product   
10030348       21553                  3  ...  High-range product   
10030349       21553                  9  ...  High-range product   
...              ...                ...  ...                 ...   
29166209       33664                  1  ...  High-range product   
29166210       33664                  1  ...  High-range product   
29166211       33664                 25  ...  High-range product   
29166212       33664                  8  ...  High-range product   
29166213       33664                 23  ...  High-range product   

             busiest_day        busiest_days  busiest_period_of_day  \
10030345  Regularly busy  Regulary busy days            Most orders   
10030346  Regularly busy  Regulary busy days            Most orders   
10030347  Regularly busy  Regulary busy days         Average orders   
10030348     Busiest day        Busiest days            Most orders   
10030349  Regularly busy        Busiest days            Most orders   
...                  ...                 ...                    ...   
29166209     Busiest day        Busiest days            Most orders   
29166210     Busiest day        Busiest days            Most orders   
29166211  Regularly busy        Slowest days            Most orders   
29166212  Regularly busy        Slowest days         Average orders   
29166213     Busiest day        Busiest days            Most orders   

          max_order      loyalty_flag   mean_price spending_flag  \
10030345         40  Regular customer   108.648299  High spender   
10030346         40  Regular customer   108.648299  High spender   
10030347          4      New customer  1154.792308  High spender   
10030348          4      New customer  1154.792308  High spender   
10030349         16  Regular customer   114.426619  High spender   
...             ...               ...          ...           ...   
29166209         39  Regular customer  1106.743956  High spender   
29166210         39  Regular c

In [32]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] >100, 'prices'] = np.nan

In [33]:
df_ords_prods_merge['prices'].max()

25.0

## Exporting the data

In [34]:
# Export new file as pkl
df_ords_prods_merge.to_pickle(os.path.join(path, 'Data','Prepared Data', 'orders_products_merged_user_flags.pkl'))